memcached-operator/pkg/controller/memcached/memcached_controller.go

```diff
@@ -109,55 +110,41 @@ func (r *ReconcileMemcached) Reconcile(request reconcile.Request) (reconcile.Res
 		return reconcile.Result{}, err
 	}
 
-	// Check if the Deployment already exists, if not create a new one
-	deployment := &appsv1.Deployment{}
-	err = r.client.Get(context.TODO(), types.NamespacedName{Name: memcached.Name, Namespace: memcached.Namespace}, deployment)
-	if err != nil && errors.IsNotFound(err) {
-		// Define a new Deployment
-		dep := r.deploymentForMemcached(memcached)
-		reqLogger.Info("Creating a new Deployment.", "Deployment.Namespace", dep.Namespace, "Deployment.Name", dep.Name)
-		err = r.client.Create(context.TODO(), dep)
-		if err != nil {
-			reqLogger.Error(err, "Failed to create new Deployment.", "Deployment.Namespace", dep.Namespace, "Deployment.Name", dep.Name)
-			return reconcile.Result{}, err
-		}
-		// Deployment created successfully - return and requeue
-		// NOTE: that the requeue is made with the purpose to provide the deployment object for the next step to ensure the deployment size is the same as the spec.
-		// Also, you could GET the deployment object again instead of requeue if you wish. See more over it here: https://godoc.org/sigs.k8s.io/controller-runtime/pkg/reconcile#Reconciler
-		return reconcile.Result{Requeue: true}, nil
-	} else if err != nil {
-		reqLogger.Error(err, "Failed to get Deployment.")
-		return reconcile.Result{}, err
-	}
+	// Update Deployment if it has changed to something other than expected. If the Deployment
+	// doesn't exist it will be created (initial creation is not really a special case).
+	deployment := &appsv1.Deployment{ObjectMeta: metav1.ObjectMeta{Name: memcached.Name, Namespace: memcached.Namespace}}
+	op, err := controllerutil.CreateOrUpdate(context.TODO(), r.client, deployment, func(ignore runtime.Object) error {
```

Comment:
If we use this approach we are suggesting for the users that they should always create or update all secondary-resources no matter if is or not required. I am right? 

So, it is not a problem for the sample, however, is it a really good practice to be adopted here and in all scenarios? Will not it send too many unnecessary requests to the API-server? Would not possible face performance issues when the project has for example 100 or 1000 CR's to reconcile? 

WDYT @joelanford @grdryn ?

Reply:
Yes, by looking its code [here](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/controller/controllerutil/controllerutil.go#L138) shows that yo you are right 👍.